src/win.c

```diff
@@ -804,15 +817,9 @@ static void win_set_shadow(session_t *ps, struct managed_win *w, bool shadow_new
 	log_debug("Updating shadow property of window %#010x (%s) to %d", w->base.id,
 	          w->name, shadow_new);
 
-	if (w->state == WSTATE_UNMAPPED) {
-		// No need to add damage or update shadow
-		// Unmapped window shouldn't have any images
-		w->shadow = shadow_new;
-		assert(!w->shadow_image);
-		assert(!w->win_image);
-		assert(win_check_flags_all(w, WIN_FLAGS_IMAGES_NONE));
-		return;
-	}
+	// We don't handle property updates of non-visible windows until they are mapped.
+	assert(w->state != WSTATE_UNMAPPED && w->state != WSTATE_DESTROYING &&
+	       w->state != WSTATE_UNMAPPING);
```

Comment:
We should add similar checks in the other functions that should only be called in the critical section (i.e. "mapped" windows). Quick search yielded `win_on_win_size_change()` and `win_update_bounding_shape()`.

Reply:
Added, thanks.